### DISCUSSION TITLE
[Question] Is the intrinsics SSE4.2 _mm_cmpistrm instruction exposed somewhere ?

### DISCUSSION THREAD

**Discussion by wolfgarbe** [dotnet/runtime#63332] [Q&A]
I am looking for the _mm_cmpistrm instruction, available on SSE4.2 processors.
( PCMPISTRM  Packed Compare Implicit Length Strings, Return Mask  https://www.felixcloutier.com/x86/pcmpistrm  )
Can't find it on the Sse42 object...

**Comment by EgorBo** [comment]
See https://github.com/dotnet/runtime/issues/957#issuecomment-802388874 (and follow up comments)
TLDR: Sse2.CompareEquals + Sse2.MoveMask or Sse41.TestZ are enough for string compare and their 256bit versions exist in AVX2

**Comment by wolfgarbe** [comment]
Thank you for the quick reply. 

I'm not doing string comparison, but intersection between two sorted 16-bit arrays using SSE4.1 intrinsics (see code below).
_mm_cmpistrm (vB , vA ,_SIDD_UWORD_OPS | _SIDD_CMP_EQUAL_ANY | _SIDD_BIT_MASK ) ; would allow to compare **any** element of vector A to **any** element of vector B (**8*8=64 comparisons**). 

As far as I know, the proposed Sse2.CompareEquals / _mm256_cmpeq_epi16[ allows only a **pairwise** comparison](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/intrinsics/intrinsics-for-intel-advanced-vector-extensions-2/intrinsics-for-compare-operations/mm256-cmpeq-epi8-16-32-64.html) of the elements of vector A and Vecor B (**8 comparisons**).

Any suggestion?

```
int32_t intersect ( uint16_t *A, size_t lengthA ,
uint16_t *B, size_t lengthB ,
uint16_t * out) {
size_t count = 0; // size of intersection
size_t i = 0 , j = 0;
int vectorlength = sizeof ( __m128i ) / sizeof ( uint16_t ) ;
__m128i vA , vB , rv , sm16 ;
while (i < lengthA ) && (j < lengthB ) {
vA = _mm_loadu_si128 (( __m128i *) &A[i]) ;
vB = _mm_loadu_si128 (( __m128i *) &B[j]) ;
rv = _mm_cmpistrm (vB , vA ,
_SIDD_UWORD_OPS | _SIDD_CMP_EQUAL_ANY | _SIDD_BIT_MASK ) ;
int r = _mm_extract_epi32 (rv , 0) ;
sm16 = _mm_load_si128 ( mask16 + r) ;
__m128i p = _mm_shuffle_epi8 (vA , sm16 ) ;
_mm_storeu_si128 (( __m128i *) ( out + count ) ,p) ;
count += _mm_popcnt_u32 (r) ;
uint16_t a_max = A[i + vectorlength - 1];
uint16_t b_max = B[j + vectorlength - 1];
if ( a_max <= b_max ) i += vectorlength ;
if ( b_max <= a_max ) j += vectorlength ;
}
return count ;
}
```
Source:  Roaring Bitmaps: Implementation of an Optimized Software Library, Daniel Lemire https://arxiv.org/pdf/1709.07821.pdf

**Comment by saucecontrol** [comment]
That's an interesting use case for `pcmpistrX` with `_SIDD_CMP_EQUAL_ANY`.  I've only ever seen it used for actual character matching, such as this example in RapidJSON:

https://github.com/dotnet/runtime/blob/0e499ac03d4eacce9ce124f454d0ed4ac2cdafcd/src/native/external/rapidjson/reader.h#L296-L305

In cases like that (small number of characters to match), similar perf can be achieved when substituting multiple compares with constant vectors that have each character broadcast to all elements -- with even better performance when extended to 256-bit vectors with AVX2.

For a full intersection of 8x8 values, there won't be an approach that can equal what `_SIDD_CMP_EQUAL_ANY` gives you, until you get access to the new AVX-512 [`vpintersectX`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#techs=AVX_512&text=VP2INTERSECTD) instructions, and even those won't substitute if you want to compare `byte` or `word` lengths.

The only real hold-up to getting those missing SSE4.2 instructions into `System.Runtime.Intrinsics` is finalizing the API shape #957.  If you have ideas around that, an API proposal would be welcome, I'm sure.  #63331 would be a good template to use.

**Reply by wolfgarbe** [reply]
Thank you very much for the detailed answer. AVX-512 vpintersectX would indeed work for my use case, but is supported only by a small number of processors and is also not yet available in C#. Probably I have to resort to writing the method in C++ ([where _mm_cmpistrm is supported by the Microsoft C++ compiler](https://docs.microsoft.com/en-us/cpp/intrinsics/x64-amd64-intrinsics-list?view=msvc-170)) and then [call it as an external library from C#](https://stackoverflow.com/questions/11425202/is-it-possible-to-call-a-c-function-from-c-net).
Unfortunately, this looks more like an ugly hack than clean code.

**Reply by saucecontrol** [reply]
FYI, I've got the ball rolling with https://github.com/dotnet/runtime/issues/68204.  Let's see if we can come up with an API shape that makes sense.
